diesel print-schema produces uncompileable schema with column named "table"

### Rust Version

1.90

### Diesel Version

2.3.4

### Diesel Features

64-column-tables, numeric, uuid, serde_json

### Database Version

PostgreSQL 17

### Operating System Version

Fedora Linux 43 (Workstation Edition) x86_64 Linux 6.17.12-300.fc43.x86_64

### What happened?

Had a column named "table" on a table in postgres, this very obviously collides with foo::table that is generated by the table! macro, but there are zero guiderails or guidelines along this error path, it just broke with a hard to decipher error (due to lower code having trait errors from the table not being setup correctly) and had no obvious resolution (now using a patchfile for this specific table)

### What did you expect to happen?

diesel print-schema should print a warning (ideally provide a mitigation on its own but "what name should it use" is bound to be a never ending war) when it encounters this guaranteed-to-fail case, and suggest using diesels patchfile support + `#[sql_name]` (or even adding a new config specifically to handle this case easier, like a column rename map), rather than leaving users in the relative dark about how to resolve this issue

### Additional details

this "bug" is due to a lack of tooling around diesels reserved names, perhaps every name that table! reserves should get a similar resolution

### Steps to reproduce

```rust
Put this in an empty new cargo bin project (with a default-features diesel) and run a check, diesel version at time of this reproducer: 2.3.5

diesel::table! {
    obfuscated (o1, o2, table, o3) {
        o1 -> Varchar,
        o2 -> Varchar,
        table -> Varchar,
        o3 -> Timestamp,
        created_at -> Timestamp,
    }
}

fn main() {
    println!("Hello, world!");
}
```

### Compile time error
As promised, its a very large unreadable error
<details>

````
    Checking dtest v0.1.0 (/home/ultrabear/programs/rust/dtest)
error[E0255]: the name `table` is defined multiple times
 --> src/main.rs:5:9
  |
1 | diesel::table! {
  | -------------- previous import of the type `table` here
...
5 |         table -> Varchar,
  |         ^^^^^ `table` redefined here
  |
  = note: `table` must be defined only once in the type namespace of this module
  = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you can use `as` to change the binding name of the import
  |
9 | } as other_table
  |   ++++++++++++++

error[E0277]: cannot select `obfuscated::table` from `obfuscated::table`
  --> src/main.rs:1:1
   |
 1 | / diesel::table! {
 2 | |     obfuscated (o1, o2, table, o3) {
 3 | |         o1 -> Varchar,
 4 | |         o2 -> Varchar,
...  |
 9 | | }
   | |_^ unsatisfied trait bound
   |
   = help: the trait `SelectableExpression<obfuscated::table>` is not implemented for `obfuscated::table`
   = note: `obfuscated::table` is no valid selection for `obfuscated::table`
   = help: the following other types implement trait `SelectableExpression<QS>`:
             `&'a T` implements `SelectableExpression<QS>`
             `(T, T1)` implements `SelectableExpression<QS>`
             `(T, T1, T2)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5, T6)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5, T6, T7)` implements `SelectableExpression<QS>`
           and 125 others
   = note: required for `(o1, o2, table, o3, created_at)` to implement `SelectableExpression<obfuscated::table>`
note: required by a bound in `diesel::QuerySource::DefaultSelection`
  --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_source/mod.rs:31:28
   |
31 |     type DefaultSelection: SelectableExpression<Self>;
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `QuerySource::DefaultSelection`
   = note: the full name for the type has been written to '/home/ultrabear/programs/rust/dtest/target/debug/deps/dtest-84d591618bb79fad.long-type-7975825408374989833.txt'
   = note: consider using `--verbose` to print the full type name to the console
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `(o1, o2, table, o3, created_at): Expression` is not satisfied
   --> src/main.rs:1:1
    |
  1 | / diesel::table! {
  2 | |     obfuscated (o1, o2, table, o3) {
  3 | |         o1 -> Varchar,
  4 | |         o2 -> Varchar,
...   |
  9 | | }
    | |_^ the trait `diesel::Expression` is not implemented for `(o1, o2, table, o3, created_at)`
    |
    = help: the following other types implement trait `diesel::Expression`:
              &T
              AliasedField<S, C>
              Box<T>
              CaseWhen<expression::case_when::non_public_types::CaseWhenConditionsIntermediateNode<W, T, Whens>, E>
              CaseWhen<expression::case_when::non_public_types::CaseWhenConditionsLeaf<W, T>, expression::case_when::non_public_types::ElseExpression<E>>
              CaseWhen<expression::case_when::non_public_types::CaseWhenConditionsLeaf<W, T>, expression::case_when::non_public_types::NoElseExpression>
              CountStar
              Exists<T>
            and 69 others
note: required by a bound in `diesel::query_builder::AsQuery::Query`
   --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_builder/mod.rs:361:23
    |
361 |     type Query: Query<SqlType = Self::SqlType>;
    |                       ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `AsQuery::Query`
    = note: the full name for the type has been written to '/home/ultrabear/programs/rust/dtest/target/debug/deps/dtest-84d591618bb79fad.long-type-7975825408374989833.txt'
    = note: consider using `--verbose` to print the full type name to the console
    = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `obfuscated::table: ValidGrouping<()>` is not satisfied
   --> src/main.rs:1:1
    |
  1 | / diesel::table! {
  2 | |     obfuscated (o1, o2, table, o3) {
  3 | |         o1 -> Varchar,
  4 | |         o2 -> Varchar,
...   |
  9 | | }
    | |_^ the trait `ValidGrouping<()>` is not implemented for `obfuscated::table`
    |
    = help: the following other types implement trait `ValidGrouping<GroupByClause>`:
              `&T` implements `ValidGrouping<GB>`
              `()` implements `ValidGrouping<GB>`
              `(T, T1)` implements `ValidGrouping<__GroupByClause>`
              `(T, T1, T2)` implements `ValidGrouping<__GroupByClause>`
              `(T, T1, T2, T3)` implements `ValidGrouping<__GroupByClause>`
              `(T, T1, T2, T3, T4)` implements `ValidGrouping<__GroupByClause>`
              `(T, T1, T2, T3, T4, T5)` implements `ValidGrouping<__GroupByClause>`
              `(T, T1, T2, T3, T4, T5, T6)` implements `ValidGrouping<__GroupByClause>`
            and 96 others
    = note: required for `(obfuscated::table, columns::o3, columns::created_at)` to implement `ValidGrouping<()>`
    = note: 2 redundant requirements hidden
    = note: required for `(o1, o2, table, o3, created_at)` to implement `ValidGrouping<()>`
    = note: required for `SelectStatement<FromClause<obfuscated::table>>` to implement `Query`
note: required by a bound in `diesel::query_builder::AsQuery::Query`
   --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_builder/mod.rs:361:23
    |
361 |     type Query: Query<SqlType = Self::SqlType>;
    |                       ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `AsQuery::Query`
    = note: the full name for the type has been written to '/home/ultrabear/programs/rust/dtest/target/debug/deps/dtest-84d591618bb79fad.long-type-7975825408374989833.txt'
    = note: consider using `--verbose` to print the full type name to the console
    = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `obfuscated::table: ValidGrouping<()>` is not satisfied
  --> src/main.rs:1:1
   |
 1 | / diesel::table! {
 2 | |     obfuscated (o1, o2, table, o3) {
 3 | |         o1 -> Varchar,
 4 | |         o2 -> Varchar,
...  |
 9 | | }
   | |_^ the trait `ValidGrouping<()>` is not implemented for `obfuscated::table`
   |
   = help: the following other types implement trait `ValidGrouping<GroupByClause>`:
             `&T` implements `ValidGrouping<GB>`
             `()` implements `ValidGrouping<GB>`
             `(T, T1)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3, T4)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3, T4, T5)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3, T4, T5, T6)` implements `ValidGrouping<__GroupByClause>`
           and 96 others
   = note: required for `(obfuscated::table, columns::o3)` to implement `ValidGrouping<()>`
   = note: 2 redundant requirements hidden
   = note: required for `(columns::o1, columns::o2, obfuscated::table, columns::o3)` to implement `ValidGrouping<()>`
note: required by a bound in `diesel::Table::PrimaryKey`
  --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_source/mod.rs:59:51
   |
59 |     type PrimaryKey: SelectableExpression<Self> + ValidGrouping<()>;
   |                                                   ^^^^^^^^^^^^^^^^^ required by this bound in `Table::PrimaryKey`
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot select `obfuscated::table` from `obfuscated::table`
  --> src/main.rs:1:1
   |
 1 | / diesel::table! {
 2 | |     obfuscated (o1, o2, table, o3) {
 3 | |         o1 -> Varchar,
 4 | |         o2 -> Varchar,
...  |
 9 | | }
   | |_^ unsatisfied trait bound
   |
   = help: the trait `SelectableExpression<obfuscated::table>` is not implemented for `obfuscated::table`
   = note: `obfuscated::table` is no valid selection for `obfuscated::table`
   = help: the following other types implement trait `SelectableExpression<QS>`:
             `&'a T` implements `SelectableExpression<QS>`
             `(T, T1)` implements `SelectableExpression<QS>`
             `(T, T1, T2)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5, T6)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5, T6, T7)` implements `SelectableExpression<QS>`
           and 125 others
   = note: required for `(columns::o1, columns::o2, obfuscated::table, columns::o3)` to implement `SelectableExpression<obfuscated::table>`
note: required by a bound in `diesel::Table::PrimaryKey`
  --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_source/mod.rs:59:22
   |
59 |     type PrimaryKey: SelectableExpression<Self> + ValidGrouping<()>;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Table::PrimaryKey`
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `obfuscated::table: ValidGrouping<()>` is not satisfied
  --> src/main.rs:1:1
   |
 1 | / diesel::table! {
 2 | |     obfuscated (o1, o2, table, o3) {
 3 | |         o1 -> Varchar,
 4 | |         o2 -> Varchar,
...  |
 9 | | }
   | |_^ the trait `ValidGrouping<()>` is not implemented for `obfuscated::table`
   |
   = help: the following other types implement trait `ValidGrouping<GroupByClause>`:
             `&T` implements `ValidGrouping<GB>`
             `()` implements `ValidGrouping<GB>`
             `(T, T1)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3, T4)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3, T4, T5)` implements `ValidGrouping<__GroupByClause>`
             `(T, T1, T2, T3, T4, T5, T6)` implements `ValidGrouping<__GroupByClause>`
           and 96 others
   = note: required for `(obfuscated::table, columns::o3, columns::created_at)` to implement `ValidGrouping<()>`
   = note: 2 redundant requirements hidden
   = note: required for `(o1, o2, table, o3, created_at)` to implement `ValidGrouping<()>`
note: required by a bound in `diesel::Table::AllColumns`
  --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_source/mod.rs:61:51
   |
61 |     type AllColumns: SelectableExpression<Self> + ValidGrouping<()>;
   |                                                   ^^^^^^^^^^^^^^^^^ required by this bound in `Table::AllColumns`
   = note: the full name for the type has been written to '/home/ultrabear/programs/rust/dtest/target/debug/deps/dtest-84d591618bb79fad.long-type-7975825408374989833.txt'
   = note: consider using `--verbose` to print the full type name to the console
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot select `obfuscated::table` from `obfuscated::table`
  --> src/main.rs:1:1
   |
 1 | / diesel::table! {
 2 | |     obfuscated (o1, o2, table, o3) {
 3 | |         o1 -> Varchar,
 4 | |         o2 -> Varchar,
...  |
 9 | | }
   | |_^ unsatisfied trait bound
   |
   = help: the trait `SelectableExpression<obfuscated::table>` is not implemented for `obfuscated::table`
   = note: `obfuscated::table` is no valid selection for `obfuscated::table`
   = help: the following other types implement trait `SelectableExpression<QS>`:
             `&'a T` implements `SelectableExpression<QS>`
             `(T, T1)` implements `SelectableExpression<QS>`
             `(T, T1, T2)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5, T6)` implements `SelectableExpression<QS>`
             `(T, T1, T2, T3, T4, T5, T6, T7)` implements `SelectableExpression<QS>`
           and 125 others
   = note: required for `(o1, o2, table, o3, created_at)` to implement `SelectableExpression<obfuscated::table>`
note: required by a bound in `diesel::Table::AllColumns`
  --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_source/mod.rs:61:22
   |
61 |     type AllColumns: SelectableExpression<Self> + ValidGrouping<()>;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Table::AllColumns`
   = note: the full name for the type has been written to '/home/ultrabear/programs/rust/dtest/target/debug/deps/dtest-84d591618bb79fad.long-type-7975825408374989833.txt'
   = note: consider using `--verbose` to print the full type name to the console
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `columns::table: QuerySource` is not satisfied
 --> src/main.rs:1:1
  |
1 | / diesel::table! {
2 | |     obfuscated (o1, o2, table, o3) {
3 | |         o1 -> Varchar,
4 | |         o2 -> Varchar,
... |
9 | | }
  | |_^ the trait `QuerySource` is not implemented for `columns::table`
  |
  = help: the following other types implement trait `QuerySource`:
            Alias<S>
            JoinOn<Join, On>
            SelectStatement<From>
            obfuscated::table
            query_source::joins::Join<Left, Right, Inner>
            query_source::joins::Join<Left, Right, LeftOuter>
  = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `walk_ast` exists for struct `StaticQueryFragmentInstance<columns::table>`, but its trait bounds were not satisfied
  --> src/main.rs:1:1
   |
 1 | / diesel::table! {
 2 | |     obfuscated (o1, o2, table, o3) {
 3 | |         o1 -> Varchar,
 4 | |         o2 -> Varchar,
 5 | |         table -> Varchar,
   | |         ----- doesn't satisfy `columns::table: StaticQueryFragment`
...  |
 9 | | }
   | | ^
   | |_|
   |   method cannot be called on `StaticQueryFragmentInstance<columns::table>` due to unsatisfied trait bounds
   |
  ::: /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_builder/nodes/mod.rs:13:1
   |
13 |   pub struct StaticQueryFragmentInstance<T>(PhantomData<T>);
   |   ----------------------------------------- doesn't satisfy `_: QueryFragment<_>`
   |
   = note: the following trait bounds were not satisfied:
           `columns::table: StaticQueryFragment`
           which is required by `StaticQueryFragmentInstance<columns::table>: QueryFragment<_>`
note: the trait `StaticQueryFragment` must be implemented
  --> /home/ultrabear/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.3.5/src/query_builder/nodes/mod.rs:6:1
   |
 6 | pub trait StaticQueryFragment {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0255, E0277, E0599.
For more information about an error, try `rustc --explain E0255`.
error: could not compile `dtest` (bin "dtest") due to 11 previous errors
````

</details>

### Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.

- [x] This issue can be reproduced on Rust's stable channel. (Please submit the issue in the Rust issue tracker instead if that's not the case)
- [x] This issue can be reproduced without requiring a third party crate

--- a/diesel_cli/src/infer_schema_internals/inference.rs
+++ b/diesel_cli/src/infer_schema_internals/inference.rs
@@ -64,6 +64,7 @@ static RESERVED_NAMES: &[&str] = &[
     "while",
     "yield",
     "bool",
+    "table",
     "columns",
     "is_nullable",
 ];

--- a/diesel_cli/tests/print_schema.rs
+++ b/diesel_cli/tests/print_schema.rs
@@ -386,6 +386,11 @@ fn print_schema_reserved_names() {
 }
 
 #[test]
+fn print_schema_column_named_table() {
+    test_print_schema("print_schema_column_named_table", vec![])
+}
+
+#[test]
 #[cfg(feature = "postgres")]
 fn print_schema_regression_3446_ignore_compound_foreign_keys() {
     test_print_schema("print_schema_regression_3446_compound_keys", vec![])

--- a/diesel_cli/tests/print_schema/print_schema_column_named_table/diesel.toml
+++ b/diesel_cli/tests/print_schema/print_schema_column_named_table/diesel.toml
@@ -1,0 +1,4 @@
+[print_schema]
+file = "src/schema.rs"
+with_docs = false
+

--- a/diesel_cli/tests/print_schema/print_schema_column_named_table/mysql/schema.sql
+++ b/diesel_cli/tests/print_schema/print_schema_column_named_table/mysql/schema.sql
@@ -1,0 +1,5 @@
+CREATE TABLE column_named_table (
+    id INTEGER PRIMARY KEY,
+    `table` INTEGER NOT NULL
+);
+

--- a/diesel_cli/tests/print_schema/print_schema_column_named_table/postgres/schema.sql
+++ b/diesel_cli/tests/print_schema/print_schema_column_named_table/postgres/schema.sql
@@ -1,0 +1,5 @@
+CREATE TABLE column_named_table (
+    id SERIAL PRIMARY KEY,
+    "table" INTEGER NOT NULL
+);
+

--- a/diesel_cli/tests/print_schema/print_schema_column_named_table/sqlite/schema.sql
+++ b/diesel_cli/tests/print_schema/print_schema_column_named_table/sqlite/schema.sql
@@ -1,0 +1,5 @@
+CREATE TABLE column_named_table (
+    id INTEGER PRIMARY KEY,
+    "table" INTEGER NOT NULL
+);
+

